Settings page

PirtleShell opened the pull request:
# A settings page for hexo-admin!
Hi there! I'm glad this is alive! I also added an extendable settings page that lets you easily change some things about the editor. I put in settings for spellchecking the editor and adding/removing line numbers (#50), but it should be easy to hook in more settings as folks see fit. I stopped keeping multiple branches for different things I was doing, so it's possible there are more changes than just adding the settings page (for instance the edits in [config-dropper.js](https://github.com/jaredly/hexo-admin/compare/master...PirtleShell:settings-page?expand=1#diff-39e59c224f79df66ce2ce10ed52c29c4)...). Apologies! Let me know if you need anything changed.

I can squash it all into one commit if that's what you're into. I'm leaving it as is for right now so you can see better what's going on. I extended the demo, too, so you can clone it and check it out. It works better in real life, but the demo is functional.

## Decide which settings you want
![hexo-admin-settings-page](https://cloud.githubusercontent.com/assets/14897503/21078650/df34a12a-bf3c-11e6-98b1-8eefbcb04119.jpg)

## Now you're editing with swagger
![edit-post-with-settings-on](https://cloud.githubusercontent.com/assets/14897503/21078654/f3e44ee0-bf3c-11e6-9858-5eeff6bf12ee.png)

PS. You write pretty code. Thanks for making this; it makes for some enjoyable writing.

jaredly commented on docs/demo/test-api.js:
```diff
@@ -69,6 +70,18 @@ module.exports = function (config) {
+    setSetting: (name, value, addOptions) => {
+      console.log(config.settings)
+      if (!config.settings.options) {
+        config.settings.options = new Object()
```
let's do `{}` instead of `new Object()`

jaredly commented on docs/demo/test-api.js:
```diff
@@ -69,6 +70,18 @@ module.exports = function (config) {
+      }
+      config.settings.options[name] = value
+      config.settings = deepAssign(config.settings, addOptions)
+      Promise.resolve({
```
gotta `return` here

jaredly commented on docs/demo/test-api.js:
```diff
@@ -69,6 +70,18 @@ module.exports = function (config) {
     },
     tagsAndCategories: () => Promise.resolve(config.tagsAndCategories),
+    settings: () => Promise.resolve(config.settings),
+    setSetting: (name, value, addOptions) => {
```
what does addOptions do? it's not clear form the usage

PirtleShell replied:
This allows a single setting to add multiple items to the settings object.

For instance, to add spellcheck to a CodeMirror instance, you need too initiate it with the options `inputStyle='content-ediatble'` and `spellcheck=true`. As an api call, setting this is
```js
api.setSetting('spellcheck', true, {editor: {inputStyle: 'content-editable', spellcheck: true}})
```
which makes `_admin-config.yml` (or `api.settings()`):
```yaml
options:
   spellcheck: true
editor:
   inputStyle: 'content-editable'
   spellcheck: true
```

The `editor` settings are [passed to CodeMirror](https://github.com/jaredly/hexo-admin/pull/91/files#diff-93bc2a7742610064cd1e94a3138c37b1R28).

The settings checkbox takes enable and disable options and passes them appropriately to `setSettings`. The checkbox has
```yaml
enableOptions: {editor: {inputStyle: 'contenteditable', spellcheck: true}}
disableOptions: {editor: {inputStyle: null, spellcheck: false}}
```
So the above is set in `_admin-config.yml` or a call to `api.settings()` when the box is enabled, and when it's diabled it looks like
```yaml
options:
   spellcheck: false
editor:
   inputStyle: null
   spellcheck: false
```

Perhaps renaming `addOptions` to `settingsToAdd` would be better?

jaredly replied:
ahh yes, how about `addedOptions`. `addOptions` felt like it was going to be a boolean or something.

PirtleShell replied:
I changed all instances accordingly.

PirtleShell commented:
Let me know if you'd like anything else changed or want it all turned into one commit.

I added some more settings I could include in this if you'd like. Or I can open a new PR when we pull this.

The additions are a settings textbox that will allow for adding single text field settings. I added the ability to customize how pasted images are handled. You can set an image directory other than `images/`, can have your own default image prefix (instead of `pasted-`) and added the ability to give a custom filename on paste. All of them optional and not affecting or changing anything if you don't change them in the settings.

Maybe better to pull this and make a new PR for the

PirtleShell commented:
Sorry, hit the wrong comment button....

PirtleShell commented:
Can we pull this, or is there more you'd like changed? I've added more cool stuff, but I built them on top of this commit, so I'm waiting to PR. I'm trying to keep your merging easy-breezy. Let me know! :grinning:

jaredly commented:
solid, thanks for sticking with this! Sorry for the delay, had holiday things
